apps/extension/src/background/approvals/service.ts

```diff
@@ -2,74 +2,162 @@ import browser from "webextension-polyfill";
+import { KeyRingService, TabStore } from "background/keyring";
+import { LedgerService } from "background/ledger";
 
 export class ApprovalsService {
   constructor(
     protected readonly txStore: KVStore<string>,
     protected readonly connectedTabsStore: KVStore<TabStore[]>,
     protected readonly keyRingService: KeyRingService,
-    protected readonly chainId: string,
-    protected readonly requester: ExtensionRequester
-  ) {}
+    protected readonly ledgerService: LedgerService
+  ) { }
 
   // Deserialize transfer details and prompt user
-  async approveTransfer(txMsg: string): Promise<void> {
+  async approveTransfer(txMsg: string, type?: AccountType): Promise<void> {
     const txMsgBuffer = Buffer.from(fromBase64(txMsg));
     const id = uuid();
-    this.txStore.set(id, txMsg);
+    await this.txStore.set(id, txMsg);
 
     // Decode tx details and launch approval screen
     const txDetails = deserialize(txMsgBuffer, TransferMsgValue);
     const { source, target, token, amount: amountBN } = txDetails;
     const amount = new BigNumber(amountBN.toString());
     const url = `${browser.runtime.getURL(
       "approvals.html"
-    )}#/tx?id=${id}&source=${source}&target=${target}&token=${token}&amount=${amount}`;
+    )}#/approve-transfer?type=${type}&id=${id}&source=${source}&target=${target}&token=${token}&amount=${amount}`;
 
-    browser.windows.create({
-      url,
-      width: 415,
-      height: 510,
-      type: "popup",
-    });
+    this._launchApprovalWindow(url);
+  }
+
+  // Deserialize bond details and prompt user
+  async approveBond(
+    txMsg: string,
+    type: AccountType,
+    publicKey?: string
+  ): Promise<void> {
+    const txMsgBuffer = Buffer.from(fromBase64(txMsg));
+    const id = uuid();
+    await this.txStore.set(id, txMsg);
+
+    // Decode tx details and launch approval screen
+    const txDetails = deserialize(txMsgBuffer, SubmitBondMsgValue);
+
+    const { source, nativeToken, amount: amountBN } = txDetails;
+    const amount = new BigNumber(amountBN.toString());
+    let url = `${browser.runtime.getURL(
+      "approvals.html"
+    )}#/approve-bond?type=${type}&id=${id}&source=${source}&token=${nativeToken}&amount=${amount}`;
 
-    return;
+    if (publicKey) {
+      url += `&publicKey=${publicKey}`;
+    }
+
+    this._launchApprovalWindow(url);
+  }
+
+  // Deserialize bond details and prompt user
+  // TODO: Finish implementing!
+  async approveUnbond(txMsg: string, type?: AccountType): Promise<void> {
+    const txMsgBuffer = Buffer.from(fromBase64(txMsg));
+    const id = uuid();
+    await this.txStore.set(id, txMsg);
+
+    // Decode tx details and launch approval screen
+    const txDetails = deserialize(txMsgBuffer, SubmitBondMsgValue);
+
+    const { source, nativeToken, amount: amountBN } = txDetails;
+    const amount = new BigNumber(amountBN.toString());
+    // TODO: This query should include perhaps a "type" indicating whether it's a bond or unbond tx:
```

Comment:
I think for `type` I will make it a part of the routing, so we can just direct it by URL to catch that param,  but the `type` I've added here is account type, not tx type - it needs to be more explicit because it's confusing!